File: 20-ValidParentheses/main.ts
const isValid = (s: string): boolean => {
  const pairs: Map<string, string> = new Map([
    [")", "("],
    ["]", "["],
    ["}", "{"],
  ]);

  const stack: string[] = [];

  for (const char of s) {
    if (pairs.get(char)) {
      if (stack.length === 0 || stack[stack.length - 1] != pairs.get(char)) {
        return false;
      }

      stack.pop();
    } else {
      stack.push(char);
    }
  }

  return stack.length === 0;
};

console.log(isValid("()"));
console.log(isValid("()[]{}"));
console.log(isValid("(]"));
console.log(isValid("([])"));

File: 20-ValidParentheses/main.py
class Solution:
  def is_valid(self, s: str) -> bool:
    stack = []
    pairs = {
      ']': '[',
      ')': '(',
      '}': '{'
    }
    
    for _, char in enumerate(s):
      if char in pairs:
        if not stack or stack[-1] != pairs[char]:
          return False
        stack.pop()
      else:
        stack.append(char)
    
    return not stack

s = Solution()
print(s.is_valid("()"))
print(s.is_valid("()[]{}"))
print(s.is_valid("(]"))
print(s.is_valid("([])"))